docs/proposals/kubectl-extension.md

```diff
@@ -0,0 +1,85 @@
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubectl Extension
+
+Abstract
+--------
+
+Allow `kubectl` to be extended to include other commands that can provide new functionality without recompiling Kubectl
+
+
+Motivation and Background
+-------------------------
+
+Kubernetes is designed to be a composable and extensible system, with the ability to add new APIs and features via Third Party Resources
+or API federation, by making the server provide functionality that eases writing generic clients, and by supporting other authentication
+systems. Given that `kubectl` is the primary method for interacting with the server, some new extensions are difficult to make usable
+for end users without recompiling that command. In addition, it is difficult to prototype new functionality for kubectl outside of the
+Kubernetes source tree.
+
+Ecosystem tools like OpenShift, Deis, and Helm add additional workflow around kubectl targeted at the end user. It is beneficial
+to encourage workflows to develop around Kubernetes without requiring them to be part of Kubernetes to both the end user community
+and the Kubernetes developer community.
+
+There are many tools that currently offer CLI extension for the same reasons - [Git](https://www.kernel.org/pub/software/scm/git/docs/howto/new-command.html) and
+[Heroku](https://devcenter.heroku.com/articles/developing-cli-plug-ins#creating-the-package) are two relevant examples in the space.
+
+
+Proposal
+--------
+
+Define a system for `kubectl` that allows new subcommands and subcommand trees to be added by placing an executable in a specific
+location on disk, like Git.  Allow third parties to extend kubectl by placing their extensions in that directory. Ensure that help
+and other logic correctly includes those extensions.
+
+A kubectl command extension would be an executable located in `EXEC_PATH` (an arbitrary directory to be defined that follows similar
+conventions in Linux) with a name pattern like `kubectl-COMMAND[-SUBCOMMAND[...]]` with one or many sub parts. The presence of
+a command extension overrides any built in command.
+
+A key requirement is that the lookup be fast (since it would be invoked on every execution of `kubectl`) and so some true extension
+behavior (such as complex inference of commands) may not be supported in order to reduce the complexity of the lookup.
+
+Kubectl would lazily include the appropriate commands in preference to the internal command structure if detected (a user asking for
+`kubectl a b c` would *first* check for `kubectl-a-b-c`, `kubectl-a-b`, or `kubectl-a` before loading the internal command).
+
+All kubectl command extensions MUST:
+
+* Support the `-h` and `--help` flags to display a help page
+* Respect the semantics of KUBECONFIG lookup (to be further specified)
+
+All kubectl command extensions SHOULD:
+
+* Follow the display and output conventions of normal kubectl commands.
```

Comment:
Thinking about something like helm - that would probably just own a
subcommand space,  I would have trouble imagining it being expressible via
anything other than complete handoff from the stub to the child proc.

On Sun, Aug 14, 2016 at 12:06 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> JSON API for what specifically?  In practice the generic object endpoints
> should be on the API server (get / describe) - the more specialized
> commands probably depend on generic server interfaces (Scale).
>
> Some of the obvious command extensions are workflow related - local
> content being transformed into Kube API objects, build integration,
> orchestration. Those are probably too complex for an API between stub and
> client.  Create commands should probably be things that can be generated
> from swagger or a similar description.  Anything that wants to exec /
> attach would be unable to rely on an request/response API.
>
>
> On Sat, Aug 13, 2016 at 11:37 PM, Brendan Burns <notifications@github.com>
> wrote:
>
>> In docs/proposals/kubectl-extension.md
>> <https://github.com/kubernetes/kubernetes/pull/30086#discussion_r74695408>
>> :
>>
>> > +a command extension overrides any built in command.
>> > +
>> > +A key requirement is that the lookup be fast (since it would be invoked on every execution of `kubectl`) and so some true extension
>> > +behavior (such as complex inference of commands) may not be supported in order to reduce the complexity of the lookup.
>> > +
>> > +Kubectl would lazily include the appropriate commands in preference to the internal command structure if detected (a user asking for
>> > +`kubectl a b c` would *first* check for `kubectl-a-b-c`, `kubectl-a-b`, or `kubectl-a` before loading the internal command).
>> > +
>> > +All kubectl command extensions MUST:
>> > +
>> > +* Support the `-h` and `--help` flags to display a help page
>> > +* Respect the semantics of KUBECONFIG lookup (to be further specified)
>> > +
>> > +All kubectl command extensions SHOULD:
>> > +
>> > +* Follow the display and output conventions of normal kubectl commands.
>>
>> Added benefits of the JSON api are that they would work if the binary is
>> local, *or* if it is deployed in the cluster as suggested by @lavalamp
>> <https://github.com/lavalamp> above.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/30086/files/b013f3367d22b4e16e3347d4e1e9ab7c2c0ea56a#r74695408>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p5zuByOKOeU_IdxTXANUqfHwos58ks5qfo1dgaJpZM4Jc6R_>
>> .
>>
>
>